ml_in_a_box.sh

```diff
@@ -1,283 +1,104 @@
-#
-# Paperspace-internal GitHub repo for this ML-in-a-Box setup: https://github.com/Paperspace/ml-in-a-box
-#
-# Last updated: Oct 04th 2021
+#!/usr/bin/env bash
 
-### TODO ###
-
-# These should be done before the VM is released
-
-# Engineering QA the GPU/CUDA/NVidia setup, especially w.r.t.
-#  - A100 GPU & CUDA so "it just works"
-#  - Versioning (most things here are fixed version, some are not; ones not have no obvious way to specify a version)
-#  - Security (we're using installed repos with arbitrary code)
-#  - Run script in one go as the superuser? (commands require sudo; VM doesn't let me be SU)
-#  - Licensing OK (looks good to me but I am not a lawyer)
-#  - Test out final VM when built
-
-# Improvements
-
-# These improve the script but are not vital
-
-# Add PyTorch 1.9/1.10 when CUDA 11.4 is supported
-# CUDA toolkit install is not verified. Later steps use it but there is a verify install using manual steps.
-# Put NumPy back to 1.21.2 after TensorFlow 2.5.0 downgrades it: pip install numpy 1.21.2 again, and then check TensorFlow still works
-# Unclear what the lib errors are when TF is asked to show the GPU devices after install
-# NVidia RAPIDS requires own virtualenv (conda, package installs, etc.), which is not ideal when nothing else installed needs it
-
-# Working directory for script
-
-mkdir ~/src
-cd ~/src
-
-
-# GPU
-# ---
-
-# - NVidia Driver 470.63.01 -
-
-# VM is set up by engineering with latest driver as of Sep 15th 2021: 470.63.01
-
-# - CUDA 11.4 -
-
-# CUDA 11.4 is present on the VM as supplied by engineering
-
-# - cuDNN 8.2.4.15-1+cuda11.4 -
-
-# Following https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html section 2.3.4.1. Ubuntu Network Installation
+sudo apt-get update
+sudo apt-get -y upgrade
+sudo apt-get -y install \
+  gcc \
+  make \
+  pkg-config \
+  apt-transport-https \
+  ca-certificates
+
+# Blacklist nouveau and rebuild kernel initramfs
+echo "blacklist nouveau
+options nouveau modeset=0" >> blacklist-nouveau.conf
+sudo mv blacklist-nouveau.conf /etc/modprobe.d/blacklist-nouveau.conf
+sudo update-initramfs -u
+sudo reboot
+
+# Install NVIDIA Linux toolkit 510.54
+wget https://us.download.nvidia.com/XFree86/Linux-x86_64/510.54/NVIDIA-Linux-x86_64-510.54.run
+chmod +x NVIDIA-Linux-x86_64-510.54.run
+sudo bash ./NVIDIA-Linux-x86_64-510.54.run
+rm NVIDIA-Linux-x86_64-510.54.run
+
+# Install CUDA toolkit 11.3 Upgrade 1 for Ubuntu 20.04 LTS
 wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/cuda-ubuntu2004.pin
-
 sudo mv cuda-ubuntu2004.pin /etc/apt/preferences.d/cuda-repository-pin-600
-sudo apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/7fa2af80.pub
-sudo add-apt-repository "deb https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/ /"
+wget https://developer.download.nvidia.com/compute/cuda/11.3.1/local_installers/cuda-repo-ubuntu2004-11-3-local_11.3.1-465.19.01-1_amd64.deb
```

Comment:
It concerns me that the NVIDIA driver version 465.19.01 is specified in this .deb that will be apt-get'ed but we run an NVIDIA provided bash script that installs 510.54 above.

Reply:
IIRC that's either the minimum supported NVIDIA driver version for CUDA 11.3, or the one that it initially shipped paired with. In either case, it doesn't actually have an impact on our setup.